File: shell/internal/storage/model/queries.go
package model

import (
	"encoding/json"
	"fmt"
	"strings"
)

// SortTerm is one entry of a structured SortSpec's terms.
type SortTerm struct {
	Column    string `json:"column"`
	Direction string `json:"direction"` // "asc" | "desc"
}

// SortSpec is src/shared/domain/queries.ts's discriminated union (sortSpecSchema). The custom
// codec is what keeps the two arms honest: marshalling never emits the other arm's key, and
// unmarshalling rejects a value that carries neither (or a malformed one).
type SortSpec struct {
	Kind  string     // "structured" | "text"
	Terms []SortTerm // structured only
	Text  string     // text only, <= 4096 (queries.ts's own cap)
}

const maxSortSpecText = 4096

func (s SortSpec) MarshalJSON() ([]byte, error) {
	switch s.Kind {
	case "structured":
		return json.Marshal(struct {
			Kind  string     `json:"kind"`
			Terms []SortTerm `json:"terms"`
		}{Kind: "structured", Terms: s.Terms})
	case "text":
		return json.Marshal(struct {
			Kind string `json:"kind"`
			Text string `json:"text"`
		}{Kind: "text", Text: s.Text})
	default:
		return nil, fmt.Errorf("model: SortSpec: invalid kind %q", s.Kind)
	}
}

func (s *SortSpec) UnmarshalJSON(data []byte) error {
	var probe struct {
		Kind  string          `json:"kind"`
		Terms json.RawMessage `json:"terms"`
		Text  json.RawMessage `json:"text"`
	}
	if err := json.Unmarshal(data, &probe); err != nil {
		return err
	}
	switch probe.Kind {
	case "structured":
		if probe.Terms == nil {
			return fmt.Errorf("model: SortSpec: structured kind missing terms")
		}
		var terms []SortTerm
		if err := json.Unmarshal(probe.Terms, &terms); err != nil {
			return fmt.Errorf("model: SortSpec: invalid terms: %w", err)
		}
		s.Kind, s.Terms, s.Text = "structured", terms, ""
		return nil
	case "text":
		if probe.Text == nil {
			return fmt.Errorf("model: SortSpec: text kind missing text")
		}
		var text string
		if err := json.Unmarshal(probe.Text, &text); err != nil {
			return fmt.Errorf("model: SortSpec: invalid text: %w", err)
		}
		if len(text) > maxSortSpecText {
			return fmt.Errorf("model: SortSpec: text exceeds %d characters", maxSortSpecText)
		}
		s.Kind, s.Terms, s.Text = "text", nil, text
		return nil
	default:
		return fmt.Errorf("model: SortSpec: unknown kind %q", probe.Kind)
	}
}

// FilterBody is queries.ts's filterBodySchema — a saved_queries row of kind 'filter'.
type FilterBody struct {
	Where   *string   `json:"where"`
	OrderBy *SortSpec `json:"orderBy"`
}

// ConsoleBody is queries.ts's consoleBodySchema — a saved_queries row of kind 'console'.
type ConsoleBody struct {
	Text string `json:"text"`
}

// SavedQuery mirrors queries.ts's savedQuerySchema. Body stays json.RawMessage rather than a Go
// sum type: the bytes are the wire value, and the repo validates them against Kind by
// unmarshalling into FilterBody/ConsoleBody and discarding the result — one struct on the wire,
// no Go-side union the renderer would have to be taught about.
type SavedQuery struct {
	ID           string          `json:"id"`
	ConnectionID string          `json:"connectionId"`
	Path         string          `json:"path"`
	Name         string          `json:"name"`
	Kind         string          `json:"kind"`
	Body         json.RawMessage `json:"body"`
	Pinned       bool            `json:"pinned"`
	CreatedAt    string          `json:"createdAt"`
	UsedAt       *string         `json:"usedAt"`
}

// SavedQueryPatch is queries.ts's updateSavedQuery patch shape.
type SavedQueryPatch struct {
	Name   *string
	Pinned *bool
}

// FilterHistoryEntry mirrors queries.ts's filterHistoryEntrySchema.
type FilterHistoryEntry struct {
	ID           string    `json:"id"`
	ConnectionID string    `json:"connectionId"`
	Path         string    `json:"path"`
	Where        *string   `json:"where"`
	OrderBy      *SortSpec `json:"orderBy"`
	UsedAt       string    `json:"usedAt"`
}

// ValidSavedQueryName mirrors queries.ts's savedQueryBase name field: trimmed, 1..120 chars.
func ValidSavedQueryName(name string) error {
	trimmed := strings.TrimSpace(name)
	if len(trimmed) < 1 {
		return fmt.Errorf("model: saved query name must not be empty")
	}
	if len(trimmed) > 120 {
		return fmt.Errorf("model: saved query name exceeds 120 characters")
	}
	return nil
}

File: packages/shared/domain/queries.ts
import { z } from 'zod';

export type SortDirection = 'asc' | 'desc';
export const sortDirectionSchema = /*#__PURE__*/ z.enum(['asc', 'desc']);

/**
 * §8.5's two sort editors, one state (D6). Clicking a grid header produces the structured
 * form and mirrors its equivalent string into the `ORDER BY` box; typing in the box switches
 * to the text form and clears the header indicators. Only the structured form is machine
 * readable enough to drive keyset pagination (D7) — a `text` sort is always `offset`.
 */
export const sortSpecSchema = /*#__PURE__*/ z.discriminatedUnion('kind', [
  /*#__PURE__*/ z.object({
    kind: z.literal('structured'),
    terms: /*#__PURE__*/ z.array(
      /*#__PURE__*/ z.object({ column: z.string(), direction: sortDirectionSchema }),
    ),
  }),
  /*#__PURE__*/ z.object({
    kind: z.literal('text'),
    text: z.string().max(4096),
  }),
]);
export type SortSpec = z.infer<typeof sortSpecSchema>;

export const filterBodySchema = /*#__PURE__*/ z.object({
  where: z.string().nullable(),
  orderBy: sortSpecSchema.nullable(),
});
export type FilterBody = z.infer<typeof filterBodySchema>;

// §8.14: "Console contents are saved to saved_queries" — just the script text, no result state.
export const consoleBodySchema = /*#__PURE__*/ z.object({
  text: z.string(),
});
export type ConsoleBody = z.infer<typeof consoleBodySchema>;

const savedQueryBase = {
  id: z.string(),
  connectionId: z.string(),
  path: z.string(),
  name: z.string().trim().min(1).max(120),
  pinned: z.boolean(),
  createdAt: z.string(),
  usedAt: z.string().nullable(),
};

export const savedQuerySchema = /*#__PURE__*/ z.discriminatedUnion('kind', [
  /*#__PURE__*/ z.object({ ...savedQueryBase, kind: z.literal('filter'), body: filterBodySchema }),
  /*#__PURE__*/ z.object({
    ...savedQueryBase,
    kind: z.literal('console'),
    body: consoleBodySchema,
  }),
]);
export type SavedQuery = z.infer<typeof savedQuerySchema>;
export type SavedFilterQuery = Extract<SavedQuery, { kind: 'filter' }>;
export type SavedConsoleQuery = Extract<SavedQuery, { kind: 'console' }>;

export const filterHistoryEntrySchema = /*#__PURE__*/ z.object({
  id: z.string(),
  connectionId: z.string(),
  path: z.string(),
  where: z.string().nullable(),
  orderBy: sortSpecSchema.nullable(),
  usedAt: z.string(),
});
export type FilterHistoryEntry = z.infer<typeof filterHistoryEntrySchema>;
